### DISCUSSION TITLE
Comparsion with DynamoDBToolbox

### DISCUSSION THREAD

**Discussion by romcok** [sensedeep/dynamodb-onetable#8] [General]
I've been thinking about the difference between your **OneTable** and that **DynamoDBToolbox**. They look pretty similar, but your library has some other extensions like migrations that are appealing. I've heard Rick Houlihan in the podcast talk to Alex DeBrie about these two libraries, and I don't understand why your library has so few watches. It looks great, but can you tell us what are the other benefits of your OneTable?

I would like to ask what it's like to use your library in TypeScript projects. This is something that DynamoDBToolbox has had from the beginning, but it's not very good, or as I would expect from a TS project.

There are currently three libraries that I have noted that are usable: OneTable, DynamoDBToolbox, and TypeDorm (which is ORM). I think a small comparison could help other people choose the right tool to work with DynamoDB.

Thank you very much 🙏

**Comment by mobsense** [comment]
Thanks for the question. We should enable GitHub discussions for these.

OneTable was extracted from a production system where it was developed and evolved for over a year. This repo is very recent and watches take time to build. We are aggressively pushing the code forward and have a suite of enhancements and tools coming like our migration CLI and a GUI to view and manage DynamoDB on a logical / entity level. 

Regarding TypeScript:  you can absolutely use OneTable with TypeScript. OneTable provides type declarations for all APIs and dynamically generates types for your schema entities and attributes. These type signatures are generated from your OneTable schema automatically!  This is huge for TypeScript users.

Differences?

I'm not sure I can give a complete comparison grid, but here are a few interesting items:

- OneTable has a more evolved schema with validations, unique attributes, default values, etc.
- OneTable has strong TypeScript supporting including automatically typed schema entities and attributes. Toolbox does not type schema entities or attributes.
- OneTable is fully validating on writes & reads. Attributes have validations and OneTable picks out the real attributes and only reads / writes these. This helps keep your database clean.
- A more refined API resulting from the year in production before release while evolving and cleaning up the API.
- Templated key management from strings or functions. This uncouples keys from attributes.
- Natural "where" clause for queries that are mapped onto conditional and filter expressions.
- Transform functions on read / write to store data in alternate formats.
- Transparent auto-follow for GSIs that project only keys to retrieve primary items. 
- Attribute packing into objects for GSIs to project multiple fields into individual attributes.
- Migrations and migration CLI. These are helpful when evolving DynamoDB designs.
- Detailed single-table metrics for CloudWatch or viewing in SenseDeep.
- Flexible API with Entity and Factory level APIs. i.e you can do:  Account.get() or Table.get('Account')
- Cross-cutting contexts for supplying attributes out-of-band. Useful for PK/SK components that come from authentication. A request auth may provide an accountId which can be provided via a context instead of the API params. This helps security and accidental security holes.
- Integrated logging
- AWS V3 support
- A high level API that allows conditional expressions to be used with get, remove, update.
- The code base is small and readable. 
- An active developer community.

**Reply by romcok** [reply]
Great, thanks for your response. I really like migration and CLI. It would be great to have type control from typescript, but it's also a bit challenging to put together all those generic types and have it perfectly typed. I think there is one problem with typing that TypeDORM also struggles with. It's 1-N relationship in STD, when you select pre-joined data and you want to return it all typed. It slightly complicates the typing of entities and it would be amazing if someone solved it somehow while maintaining DynamoDB best practices.

**Reply by mobsense** [reply]
We've just committed a TypeScript update that provides typing for the schema. This is for folks to test and get feedback.

TypeScript is pretty amazing how it can generate types from schema data. 

See the issue: #9 

You can now do things like:

```
import {Model} from 'dynamodb-onetable'

const schema = {
    models: {
        Account: {
            id: {type: String},
            name: {type: String, required: true},
            age: {type: Number, default: 23},
        },
    }
}

// Schema is fully type checked
const table = new Table({ name: 'MyTable', client, schema })

type Account = OneEntity<typeof schema.models.Account>

let account: Account = {
    // Fully type checked
    id: '1234',
    name: 'Peter Smith',
    age: 33,
}

let AccountModel = new Model<Account>(table, 'Account');

// The parameters to get and the result account are fully type checked
account = AccountModel.get({id: '1234'})
```

**Comment by mobsense** [comment]
Can you explain a bit more.

Can you explain more about "generic types"?  I'm not a TypeScript expert, but I would have thought generating the definitions for each entity would then fully type the return values and parameters for the per-entity methods.

e.g.
```
let account = await Account.get({id})
```

the table level methods would be untyped:

```
let account = await table.get('Account', {id})
```

**Reply by mobsense** [reply]
No sweat, I get it now:

See issue: #9

**Reply by mobsense** [reply]
To close this out ... OneTable now has such automatic typescript typing of your database entities.

**Comment by mkovel** [comment]
Hi,
I agree with @romcok. Guys, you did a great job. Thanks. 
I started figuring out with your package and found out that I scared to use this in production because it has no tests. 
If you add tests with enough covering, I assume people will be braver to use it in production.  )

Good luck guys, I like your style `pk:          { value: 'account:${name}' },`

**Reply by mobsense** [reply]
Thank you for the feedback.

We have tests, but not yet ready for publishing. We've used this library in production for over a year in our SenseDeep developer studio. We then extracted it and published as a library. Our tests are somewhat integrated with our other code, so it is taking a while to isolate. Should not be long before we can add them to the repo.

**Reply by mkovel** [reply]
awesome, looking  forward to it
:+1:

**Reply by mkovel** [reply]
@mobsense  Could you please take a look at my PR's?

**Reply by mobsense** [reply]
Already have. Thank you for submitting, I will merge early these next week.

**Reply by mobsense** [reply]
Have a look now @mkovel, we've checked in some of our tests with reasonable test coverage ~90%.

**Comment by evbo** [comment]
I think there's another difference. onetable offers visualization capability of the table keys/indexes. There appears to be a deepsense paid for option but also Rick H's free tool here:
https://rh-web-bucket.s3.amazonaws.com/index.html#

To me, visually understanding these things is paramount. I wish Rick's or somebody's tool was also open source. Without being able to see the end product with some dummy data I get lost so easily.

Layers of abstraction and slick APIs are great, but it's the big picture and self-documentation I think of as dynamodb's biggest cognitive obstacle.

Interested to hear about what people think of better ways to visualize. Looks like Dynamodbtoolbox don't got it, so I'm not interested... (yet)
